src/wh_server.c

```diff
@@ -21,29 +21,35 @@
 #include "wolfhsm/wh_error.h"
 #include "wolfhsm/wh_comm.h"
 
-#if 0
-#include "wolfhsm/nvm.h"
-#include "wolfhsm/nvm_remote.h"
-#endif
+#include "wolfhsm/wh_nvm.h"
 
 #include "wolfhsm/wh_message.h"
 #include "wolfhsm/wh_message_comm.h"
+#include "wolfhsm/wh_message_nvm.h"
 #include "wolfhsm/wh_server.h"
 
 int wh_Server_Init(whServerContext* server, whServerConfig* config)
 {
     int rc = 0;
-    if (server == NULL) {
+    if ((server == NULL) || (config == NULL)) {
         return WH_ERROR_BADARGS;
     }
 
     memset(server, 0, sizeof(*server));
-    if (
-/*            ((rc = wh_Nvm_Init(server->nvm_device, config->nvm_device)) == 0) && */
-        ((rc = wh_CommServer_Init(server->comm, config->comm)) == 0) &&
-/*        ((rc = wh_NvmServer_Init(server->nvm, config->nvm)) == 0)*/
-        1) {
-        /* All good */
+
+    if (    (config->nvm_config->cb != NULL) &&
+            (config->nvm_config->cb->Init != NULL)) {
+
+        rc = config->nvm_config->cb->Init(config->nvm_config->context, config->nvm_config->config);
+    }
+    if (rc == 0) {
+        server->nvm->cb = config->nvm_config->cb;
+        server->nvm->context = config->nvm_config->context;
+
+        rc = wh_CommServer_Init(server->comm, config->comm_config);
+        if (rc == 0) {
+            /* All good */
+        }
```

Comment:
Not insane at all.  There's probably no downside to providing a NULL CommServer implementation so that this would still return success and any function call would return aborted (or not ready).  I think it is vital to error out of the server init if any of the components fail to initialize, but the user is free to put in really dumb implementations if they want to.  I can imagine a server build that does not support external comms an is only there to support HSM core debugging/benchmarking or for failure analysis.  I wanted to have "skeleton" implementations for each of the components, so let's use those as the NULL instances that always return reasonably to support unit testing.

Reply:
OK so we want all components to be required, (e.g. NVM or comms cant be null) but tell customers to register dummy handlers if they don't need? I think that is fine. Def don't need to do now, but if that is the case, maybe we should probably just provide a dummy handler for each layer to make that easier for them to just register if they don't want a component.